File: tools/semantic_ir/expressions.py
"""STUNIR Semantic IR Expression Nodes.

DO-178C Level A Compliant
Expression node definitions.
"""

from typing import List, Optional, Union
from pydantic import BaseModel, Field

from .ir_types import (
    NodeID, IRName, IRNodeKind,
    BinaryOperator, UnaryOperator
)
from .nodes import IRNodeBase, TypeReference


class ExpressionNode(IRNodeBase):
    """Base expression node."""
    type: TypeReference  # Required for expressions


class IntegerLiteral(ExpressionNode):
    """Integer literal expression."""
    kind: IRNodeKind = IRNodeKind.INTEGER_LITERAL
    value: int
    radix: int = Field(default=10, ge=2, le=16)


class FloatLiteral(ExpressionNode):
    """Float literal expression."""
    kind: IRNodeKind = IRNodeKind.FLOAT_LITERAL
    value: float


class StringLiteral(ExpressionNode):
    """String literal expression."""
    kind: IRNodeKind = IRNodeKind.STRING_LITERAL
    value: str


class BoolLiteral(ExpressionNode):
    """Boolean literal expression."""
    kind: IRNodeKind = IRNodeKind.BOOL_LITERAL
    value: bool


class VarRef(ExpressionNode):
    """Variable reference expression."""
    kind: IRNodeKind = IRNodeKind.VAR_REF
    name: IRName
    binding: Optional[NodeID] = None


class BinaryExpr(ExpressionNode):
    """Binary operation expression."""
    kind: IRNodeKind = IRNodeKind.BINARY_EXPR
    op: BinaryOperator
    left: Union[NodeID, "ExpressionNode"]
    right: Union[NodeID, "ExpressionNode"]


class UnaryExpr(ExpressionNode):
    """Unary operation expression."""
    kind: IRNodeKind = IRNodeKind.UNARY_EXPR
    op: UnaryOperator
    operand: Union[NodeID, "ExpressionNode"]


class FunctionCall(ExpressionNode):
    """Function call expression."""
    kind: IRNodeKind = IRNodeKind.FUNCTION_CALL
    function: Union[IRName, NodeID]
    arguments: List[Union[NodeID, "ExpressionNode"]] = Field(default_factory=list)


class MemberExpr(ExpressionNode):
    """Member access expression."""
    kind: IRNodeKind = IRNodeKind.MEMBER_EXPR
    object: Union[NodeID, "ExpressionNode"]
    member: IRName
    is_arrow: bool = False


class ArrayAccess(ExpressionNode):
    """Array access expression."""
    kind: IRNodeKind = IRNodeKind.ARRAY_ACCESS
    array: Union[NodeID, "ExpressionNode"]
    index: Union[NodeID, "ExpressionNode"]


class CastExpr(ExpressionNode):
    """Cast expression."""
    kind: IRNodeKind = IRNodeKind.CAST_EXPR
    operand: Union[NodeID, "ExpressionNode"]
    target_type: TypeReference


class TernaryExpr(ExpressionNode):
    """Ternary conditional expression."""
    kind: IRNodeKind = IRNodeKind.TERNARY_EXPR
    condition: Union[NodeID, "ExpressionNode"]
    then_expr: Union[NodeID, "ExpressionNode"]
    else_expr: Union[NodeID, "ExpressionNode"]


# Enable forward references
BinaryExpr.model_rebuild()
UnaryExpr.model_rebuild()
FunctionCall.model_rebuild()
MemberExpr.model_rebuild()
ArrayAccess.model_rebuild()
CastExpr.model_rebuild()
TernaryExpr.model_rebuild()

File: tools/rust/semantic_ir/src/expressions.rs
//! STUNIR Semantic IR Expression Nodes
//!
//! Expression node definitions.

use serde::{Deserialize, Serialize};

use crate::types::*;
use crate::nodes::*;

/// Expression node (contains type information)
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct ExpressionNode {
    #[serde(flatten)]
    pub base: IRNodeBase,
    #[serde(rename = "type")]
    pub expr_type: TypeReference,
}

/// Integer literal
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct IntegerLiteral {
    #[serde(flatten)]
    pub base: ExpressionNode,
    pub value: i64,
    #[serde(default = "default_radix")]
    pub radix: u8,
}

fn default_radix() -> u8 { 10 }

/// Float literal
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct FloatLiteral {
    #[serde(flatten)]
    pub base: ExpressionNode,
    pub value: f64,
}

/// String literal
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct StringLiteral {
    #[serde(flatten)]
    pub base: ExpressionNode,
    pub value: String,
}

/// Boolean literal
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct BoolLiteral {
    #[serde(flatten)]
    pub base: ExpressionNode,
    pub value: bool,
}

/// Variable reference
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct VarRef {
    #[serde(flatten)]
    pub base: ExpressionNode,
    pub name: IRName,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub binding: Option<NodeID>,
}

/// Binary expression
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct BinaryExpr {
    #[serde(flatten)]
    pub base: ExpressionNode,
    pub op: BinaryOperator,
    pub left: NodeID,
    pub right: NodeID,
}

/// Unary expression
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct UnaryExpr {
    #[serde(flatten)]
    pub base: ExpressionNode,
    pub op: UnaryOperator,
    pub operand: NodeID,
}

/// Function call
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct FunctionCall {
    #[serde(flatten)]
    pub base: ExpressionNode,
    pub function: IRName,
    pub arguments: Vec<NodeID>,
}

/// Member access
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct MemberExpr {
    #[serde(flatten)]
    pub base: ExpressionNode,
    pub object: NodeID,
    pub member: IRName,
    #[serde(default)]
    pub is_arrow: bool,
}

/// Array access
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct ArrayAccess {
    #[serde(flatten)]
    pub base: ExpressionNode,
    pub array: NodeID,
    pub index: NodeID,
}
